### DISCUSSION TITLE
Disconnection does not fire event until I try to send via the client

### DISCUSSION THREAD

**Discussion by EGirardi** [jchristn/SuperSimpleTcp#138] [General]
When I shut down the server side, my application using SuperSimpleTCP Client does not fire the disconnect event until i try to send something to the server - at that point it will fire the Disconnected event. How can i get the Disconnect event to fire as soon as the client gets disconnected ?

**Comment by jchristn** [comment]
Interesting, how are you shutting down the server?  Could you let me know what operating system, runtime/framework you are using? 

I just tested using the test client and server apps, hitting CTRL-C even in the server (left side) caused the client (right side) to detect the disconnect immediately.

![image](https://user-images.githubusercontent.com/1421081/169321121-8e3e95a7-4cac-4ed1-925f-079a4f8c54de.png)

**Reply by EGirardi** [reply]
Using Visual Studio 2022, .net 4.8 c#
Running on Windows 8.1 (my dev laptop).

Client connects to a embedded hardware via TCP/IP on port 555 - it's embedded hardware and the TCPIP server is always running. I manually rebooted the device to test the disconnect.

**Reply by EGirardi** [reply]
For me it only detects the disconnect once I try to send a message to the server - then it immediately detects it before trying to do the send.

**Reply by jchristn** [reply]
Got it.  So the client end is from this library, the server end is a hardware device (not using this library).

Best I can guess is that the hardware device is killing the network interface rather than (gracefully) tearing down TCP connections (i.e. sending FIN and RST).  In these cases, you might want to try enabling TCP keepalives (```client.Keepalive.EnableTcpKeepAlives = true```).  I generally don't recommend enabling them, because 1) they are finicky and 2) they are managed by the underlying OS, so the library really has limited control/influence over them.

**Reply by EGirardi** [reply]
I thought the keep alive is enabled by default (according to the GitHub page)

**Reply by jchristn** [reply]
Great catch.  The documentation is incorrect, I will fix that.  They are disabled by default: https://github.com/jchristn/SuperSimpleTcp/blob/041b57bbf1d9578341d11055b09602cc4fe72eab/src/SuperSimpleTcp/SimpleTcpKeepaliveSettings.cs#L21

**Reply by EGirardi** [reply]
Yeah - tried the keep alive - still does not detect it. I may have to put a timer in as a heartbeat that tries to send a heartbeat packet every 10 seconds or so to force it to detect the disconnect.

**Reply by EGirardi** [reply]
Seems like the embedded hardware does not respond to the keep-alive which causes the Client to terminate the connection - even though i can clearly see the client is connected in the embedded hardware side. Had to remove it from the code

**Reply by jchristn** [reply]
May I ask what the device is?

**Reply by EGirardi** [reply]
It's a Crestron Control Processor - industrial type automation controller.

**Reply by jchristn** [reply]
Got it.  Will the hardware acknowledge a message that it cannot decipher?  Or does it natively have a heartbeat function that you could take advantage of?

**Reply by EGirardi** [reply]
Its just a raw TCP connection my code will handle the heartbeat character. I did a simple test and it seems to work.
